clippy_lints/src/utils/mod.rs

```diff
@@ -330,6 +330,17 @@ pub fn method_chain_args<'a>(expr: &'a Expr, methods: &[&str]) -> Option<Vec<&'a
     Some(matched)
 }
 
+/// Returns true if the function is an entrypoint to a program
+///
+/// This is either the usual `main` function or a custom function with the `#[start]` attribute.
```

Comment:
The current impl will conservatively treat `mod foo { fn main() {} }` as the main function, too. I'm not sure if that actually works, but the docs show that we have the [entry_fn](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.entry_fn) that should give us the `DefId` of the entrypoint. Passing [LOCAL_CRATE](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/def_id/constant.LOCAL_CRATE.html) as the argument should do the trick. If that fails, there's also the way miri does it: https://github.com/solson/miri/blob/master/src/bin/miri.rs#L116